src/v/raft/replicate_batcher.cc

```diff
@@ -124,35 +145,42 @@ replicate_batcher::do_cache_with_backpressure(
      * When batch size exceed available semaphore units we just acquire all of
      * them to be able to continue.
      */
+    ssx::semaphore_units u;
+    if (timeout) {
+        u = co_await ss::get_units(
+          _max_batch_size_sem,
+          std::min(bytes, _max_batch_size),
+          ssx::semaphore::clock::now() + *timeout);
+    } else {
+        u = co_await ss::get_units(
+          _max_batch_size_sem, std::min(bytes, _max_batch_size));
+    }
 
-    return ss::get_units(_max_batch_size_sem, std::min(bytes, _max_batch_size))
-      .then(
-        [this, expected_term, batches = std::move(batches), consistency_lvl](
-          ssx::semaphore_units u) mutable {
-            size_t record_count = 0;
-            auto i = ss::make_lw_shared<item>();
-            for (auto& b : batches) {
-                record_count += b.record_count();
-                if (b.header().ctx.owner_shard == ss::this_shard_id()) {
-                    i->data.push_back(std::move(b));
-                } else {
-                    i->data.push_back(b.copy());
-                }
-            }
-            i->expected_term = expected_term;
-            i->record_count = record_count;
-            i->units = std::move(u);
-            i->consistency_lvl = consistency_lvl;
+    size_t record_count = 0;
+    std::vector<model::record_batch> local_batches;
+    local_batches.reserve(batches.size());
+    for (auto& b : batches) {
+        record_count += b.record_count();
+        if (b.header().ctx.owner_shard == ss::this_shard_id()) {
+            local_batches.push_back(std::move(b));
+        } else {
+            local_batches.push_back(b.copy());
```

Comment:
I wonder if theres a way to use foreign ptr instead of copying each batch? My hunch is that would require some bigger modifications though, right?